[E-Documents Core] - OIOUBL processing using new import process

AndriusAndrulevicius opened the pull request:
This pull request does not have a related issue as it's part of the delivery for development agreed directly with @altotovi

Implements comprehensive OIOUBL (Danish UBL) format handling for electronic document processing.

Introduces structured format reader for parsing OIOUBL invoices, credit notes, and reminders into draft purchase documents. Supports multiple vendor lookup strategies including VAT registration number, GLN, and participant identifier matching.

Extends document type enumeration to support purchase credit memos and issued reminders with appropriate draft finishing implementations.

Enables Danish businesses to process incoming electronic documents in the standardized OIOUBL format through the existing e-document framework.

> [!IMPORTANT]
> This Pull request depends on   #32

Fixes #

AndriusAndrulevicius committed:
Adds OIOUBL electronic document format support

Implements comprehensive OIOUBL (Danish UBL) format handling for electronic document processing.

Introduces structured format reader for parsing OIOUBL invoices, credit notes, and reminders into draft purchase documents. Supports multiple vendor lookup strategies including VAT registration number, GLN, and participant identifier matching.

Extends document type enumeration to support purchase credit memos and issued reminders with appropriate draft finishing implementations.

Enables Danish businesses to process incoming electronic documents in the standardized OIOUBL format through the existing e-document framework.

AndriusAndrulevicius committed:
merge main into dev/aan/OIOUBL-v2

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+    var
+        EDocumentImportHelper: Codeunit "E-Document Import Helper";
+
+    /// <summary>
+    /// Reads an OIOUBL electronic document into a draft purchase document.
+    /// Parses XML content and populates purchase header and lines based on document type (Invoice, Credit Note, or Reminder).
+    /// </summary>
+    /// <param name="EDocument">The E-Document record to process.</param>
+    /// <param name="TempBlob">The temporary blob containing the XML content to parse.</param>
+    /// <returns>Returns the process draft type indicating a Purchase Document was created.</returns>
```
Please align this documentation with documentation on PINT A-NZ format 

<img width="599" height="190" alt="Image" src="https://github.com/user-attachments/assets/6b03d26e-8726-4512-8712-eff619c26bb1" />

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+        OIOUBLXml: XmlDocument;
+        XmlNamespaces: XmlNamespaceManager;
+        XmlElement: XmlElement;
+        CommonAggregateComponentsLbl: Label 'urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2';
+        CommonBasicComponentsLbl: Label 'urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2';
+        DefaultInvoiceLbl: Label 'urn:oasis:names:specification:ubl:schema:xsd:Invoice-2';
+        DefaultCreditNoteLbl: Label 'urn:oasis:names:specification:ubl:schema:xsd:CreditNote-2';
+        DefaultReminderLbl: Label 'urn:oasis:names:specification:ubl:schema:xsd:Reminder-2';
```
Please use Tok suffix and lock these values

<img width="599" height="190" alt="Image" src="https://github.com/user-attachments/assets/3e3e83fc-4ae9-4ce3-80ca-3d497e0e99c1" />

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+        DefaultInvoiceLbl: Label 'urn:oasis:names:specification:ubl:schema:xsd:Invoice-2';
+        DefaultCreditNoteLbl: Label 'urn:oasis:names:specification:ubl:schema:xsd:CreditNote-2';
+        DefaultReminderLbl: Label 'urn:oasis:names:specification:ubl:schema:xsd:Reminder-2';
+        InvoiceLbl: Label 'INVOICE', Locked = true;
+        CreditNoteLbl: Label 'CREDITNOTE', Locked = true;
+        ReminderLbl: Label 'REMINDER', Locked = true;
```
We should stay consistent with our implementation in PINT A-NZ we left this values hardcoded:

<img width="549" height="136" alt="Image" src="https://github.com/user-attachments/assets/15aa3ebd-8b42-449a-b7e5-72672815f667" />

So either we leave them hardcoded or lets move in to the labels in both PR's. Also, the best would be here to follow Peppol example MS build if they were hardcoded there lets have them hardcoded in our solutions

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+        exit(Enum::"E-Doc. Process Draft"::"Purchase Document");
+    end;
+
+    /// <summary>
+    /// Opens a page to view the readable purchase document content for the specified E-Document.
+    /// Displays purchase header and line information in a user-friendly format.
+    /// </summary>
+    /// <param name="EDocument">The E-Document record to view.</param>
+    /// <param name="TempBlob">The temporary blob containing the document content (not used in current implementation).</param>
```
Same lets align with PINT A-NZ format documentation

<img width="645" height="115" alt="Image" src="https://github.com/user-attachments/assets/43fadbc8-7ed1-4666-b7e1-8f71b6462f7b" />

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+        // Parse parties
+        VendorNo := ParseAccountingSupplierParty(OIOUBLXml, XmlNamespaces, EDocumentPurchaseHeader, EDocument, 'inv:Invoice');
+        ParseAccountingCustomerParty(OIOUBLXml, XmlNamespaces, EDocumentPurchaseHeader, 'inv:Invoice');
+
```
```suggestion
```

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+        EDocumentXMLHelper: Codeunit "EDocument XML Helper";
+        VendorNo: Code[20];
+    begin
+        EDocumentPurchaseHeader."E-Document Type" := "E-Document Type"::"Purchase Invoice";
```
I'm curious in this format  on procedures PopulateEDocumentForInvoice(), PopulateEDocumentForCreditNote() we set document type, but lets say on PINT A-NZ we don't why it is so ?

AndriusAndrulevicius replied:
bug in PINT A-NZ

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+        // Parse parties
+        VendorNo := ParseAccountingSupplierParty(OIOUBLXml, XmlNamespaces, EDocumentPurchaseHeader, EDocument, 'cn:CreditNote');
+        ParseAccountingCustomerParty(OIOUBLXml, XmlNamespaces, EDocumentPurchaseHeader, 'cn:CreditNote');
+
```
```suggestion
```

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+        // Parse parties
+        VendorNo := ParseAccountingSupplierParty(OIOUBLXml, XmlNamespaces, EDocumentPurchaseHeader, EDocument, 'rem:Reminder');
+        ParseAccountingCustomerParty(OIOUBLXml, XmlNamespaces, EDocumentPurchaseHeader, 'rem:Reminder');
+
```
```suggestion
```

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+    begin
+        EDocumentPurchaseHeader."E-Document Type" := "E-Document Type"::"Issued Reminder";
+#pragma warning disable AA0139 // false positive: overflow handled by SetStringValueInField
+        // Basic document information
```
Those comments look like copilot generated and not really needed

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+#pragma warning disable AA0139 // false positive: overflow handled by SetStringValueInField
+        BasePathTxt := '/' + DocumentType + '/cac:AccountingSupplierParty/cac:Party';
+
+        // Extract basic vendor information
```
Same here: Those comments look like copilot generated and not really needed

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+            SchemeID := XMLNode.AsXmlAttribute().Value();
+            EndpointID := EDocumentXMLHelper.GetNodeValue(OIOUBLXml, XmlNamespaces, BasePathTxt + '/cbc:EndpointID');
+            case SchemeID of
+                'DK:CVR':
```
hardcode, move to the label

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+        VATRegistrationNo := CopyStr(EDocumentPurchaseHeader."Vendor VAT Id", 1, 20);
+        VendorName := EDocumentPurchaseHeader."Vendor Company Name";
+        VendorAddress := EDocumentPurchaseHeader."Vendor Address";
+
```
```suggestion
```

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+            SchemeID := XMLNode.AsXmlAttribute().Value();
+            EndpointID := EDocumentXMLHelper.GetNodeValue(OIOUBLXml, XmlNamespaces, BasePathTxt + '/cbc:EndpointID');
+
+            if SchemeID = 'GLN' then
```
Move GLN value to locked label

AndriusAndrulevicius committed:
Update Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Delete Apps/W1/EDocument/app/src/Document/EDocumentType.Enum.al

AndriusAndrulevicius committed:
Update Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Refactors OIOUBL handler with new enum and cleanup

Introduces new E-Document Type enum with extensible implementation support and proper interface assignments for purchase invoice and reminder processing.

Improves code maintainability by converting label variables to token constants with proper naming conventions and removes redundant comments throughout the XML parsing logic.

Adds support for credit note billing reference extraction and consolidates duplicate XML processing code for better readability.

GMatuleviciute commented on Apps/DK/EDocumentFormatOIOUBL/app/src/EDocumentOIOUBLHandler.Codeunit.al:
```diff
@@ -0,0 +1,447 @@
+                VendorNo := EDocumentImportHelper.FindVendorByNameAndAddress(VendorName, VendorAddress);
+    end;
+
+    local procedure ParseAccountingCustomerParty(OIOUBLXml: XmlDocument; XmlNamespaces: XmlNamespaceManager; var EDocumentPurchaseHeader: Record "E-Document Purchase Header"; DocumentType: Text)
```
That regarding left comments and empty lines. You already fixed it without understanding it :D

AndriusAndrulevicius committed:
Localizes helper and prevents VAT truncation

Scopes the import helper to the parsing routine to reduce global state and tighten dependencies.

Replaces a hardcoded length with MaxStrLen when copying the VAT identifier, avoiding magic numbers and preventing unintended truncation if the field size changes.

AndriusAndrulevicius committed:
Merge remote-tracking branch 'origin/main' into dev/aan/OIOUBL-v2

AndriusAndrulevicius committed:
Adds OIOUBL tests and draft reset

Introduces an integration test suite that validates end-to-end import of OIOUBL invoices into draft and finalized purchase documents, including UI-driven draft edits. Adds reusable XML parsing utilities (currency, string, number, date via XPath) to simplify readers and correctly handle non-LCY currency extraction. Adds a draft reset routine to clear purchase drafts linked to an e-document for clean retries.

Improves coverage and robustness of structured import, reduces parser duplication, and supports consistent behavior across localizations.
